### PR TITLE
Feature 914 support sierra experience events

### PR DESCRIPTION
**Pull request by sawhneyh**
PR raised for Issue: https://github.com/adobe/xdm/issues/914

### CODE REVIEW

**Comment by sawhneyh** [comment]
I have run the lint locally, and there were whitesapces issues in the existing files. The build is failing for the lint error on a file, which is not part of this PR.

File: extensions/adobe/experience/adcloud/searchads/account.example.1.json
Comment (sawhneyh) [inline extensions/adobe/experience/adcloud/searchads/account.example.1.json:9]: Existing lint issue, prior to this PR.

File: extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json
Comment (sawhneyh) [inline extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:8]: All the schema extensions are created under the hierarchy "adobe/experience/customerJourneyManagement" but the "id" field has "xdm" in path. This was require, as per the test scripts the id field has to follow the regex:
`should match pattern "(https://ns\.adobe\.com/xdm/[a-z0-9-/]*)|(http://schema\.org/.*)|(http://ns.adobe.com/adobecloud/core/1.0.*)|([http|https]://ns\.adobe\.com/adobecloudplatform/.*)|(https://tools\.ietf\.org/html/draft-kelly-json-hal-08/.*)|(http://www.iptc\.org/.*)|(https://id3\.org/id3v2.4/.*)`

**Comment by sawhneyh** [comment]
The current test scripts, doesn't verify the schemas and examples under the "/extension" folder. For testing the schema i have to modify the following scripts to validate the schemas and example added with this script. We should have separate issue for fixing the following scripts.

1. test/example-valid.js
2. test/example-presence.js
3. test/invalid-presence.js

File: extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json
Comment (sawhneyh) [inline extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json:40]: Should we have dedicated field like JourneyId for referring to the external application which initiated the message execution. Or keep it generic with field 'xdm:externalApplicationID'.

File: extensions/adobe/experience/customerJourneyManagement/message-delivery-feedback.schema.json
Comment (jfait) [inline extensions/adobe/experience/customerJourneyManagement/message-delivery-feedback.schema.json:69]: temporary

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.1.json
Comment (jfait) [inline extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.1.json:23]: do we need this given xdm:channelConext further below

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.4.json
Comment (jfait) [inline extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.4.json:1]: should we name the examples to indicate;
EVENT_TYPE & CHANNEL & STATUS

this one looks like feedback-push-exclude example for instance

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.5.json
Comment (jfait) [inline extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.5.json:1]: should we name the examples to indicate;
EVENT_TYPE & CHANNEL & STATUS

this one looks like tracking-email-click example for instance

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.5.json
Comment (jfait) [inline extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.5.json:16]: Do need message label?
Can labels on a message change overtime? any impact?

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.6.json
Comment (jfait) [inline extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.6.json:36]: Are there action details missing here?

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.6.json
Comment (jfait) [inline extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.6.json:23]: this looks like a push event but channel is email

File: extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json
Comment (jfait) [inline extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json:36]: Do we need label? is this for user friendly name display in reporting or is there any other reason to carry forward labels?

File: extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json
Comment (jfait) [reply extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json:40]: Could the ID be namespaced?
IIRC XDM had a datasource or similar that had notion of NS+ID

File: extensions/adobe/experience/customerJourneyManagement/messageprofile.schema.json
Comment (jfait) [inline extensions/adobe/experience/customerJourneyManagement/messageprofile.schema.json:31]: Can variant be more complex and nested?

What if message header had 4 variants, message body had 2, and message footer had 3.

Is that possible or would variants be a cumulative sum of outcomes?

File: extensions/adobe/experience/customerJourneyManagement/push-channel-context.example.1.json
Comment (jfait) [inline extensions/adobe/experience/customerJourneyManagement/push-channel-context.example.1.json:1]: Mobile team should also review if they haven't already

File: extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json
Comment (ctrubian) [reply extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json:36]: +1 what is this label and where does it come from?

File: extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json
Comment (ctrubian) [inline extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json:32]: Nit picky, but can we change this to "Tracker link ID"

File: extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json
Comment (ctrubian) [inline extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json:42]: "Link Category"

File: extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json
Comment (ctrubian) [reply extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json:40]: the combination of externalAppicationID and externalApplication provides a namespaced ID

Comment (ctrubian) [reply extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json:40]: and I prefer the generic approach rather than a journey specific one

**Comment by ctrubian** [comment]
In general this looks good to me for tracker purposes.  Just spelling/wording and removal or explanation of what the link label is.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.4.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.4.json:1]: The current automation scripts which validates the examples with the corresponding schema file requires, example files to have a naming convention : schema-name.example.[0-9].json. Check line:
1. https://github.com/adobe/xdm/blob/master/test/example-valid.js#L13
2. https://github.com/adobe/xdm/blob/master/test/example-valid.js#L67

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.5.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.5.json:1]: The current automation scripts which validates the examples with the corresponding schema file requires, example files to have a naming convention : schema-name.example.[0-9].json

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.1.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.1.json:23]: In xdm:channelConext we don't have any channelType field to specify for which channel is this EE for.  The consumer of this schema has to rely on presence of object like "xdm:emailChannelContext" or "xdm:pushChannelContext" to identify whether this messages is for Email or Push channel.

Though we can add an explicit channel type field in "xdm:channelConext", but for completeness of messageProfile mixin which can be used in other schemas(if required) also, I added this property here.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.5.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.5.json:16]: I added this for the use cases where customer will be building the segmentation/reporting use cases in AEP/CJA. For usability purpose i added the label fields, otherwise it will be difficult for end user to build segments/reports using the machine generated ids.

File: extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json:36]: > is this for user friendly name display in reporting or is there any other reason to carry forward labels?

Yes this is for reporting use cases in CJA.

> what is this label and where does it come from?

Enrichment service will enrich this using the link id.

File: extensions/adobe/experience/customerJourneyManagement/messageprofile.schema.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/messageprofile.schema.json:31]: I am not aware of this requirement, if we have this requirement then we should add a support for defining multiple variants with respective placement.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.6.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.6.json:36]: There are no specific requirement provided for the reporting or segmentation use cases on the metadata required with the "action" event.
Though will still check with Pdms for an explicit requirements.
cc: @aashishjassal

File: extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json
Comment (ctrubian) [reply extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json:36]: So this is something that will be set in the authoring environment and that will propagate it to somewhere that enrichment can find it?

File: extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json
Comment (cmathis) [inline extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:20]: Because this is not a core standard XDM mixin, these fields should not be in the XDM namespace either. This field should be named "https://ns.adobe.com/xdm/experience/customerJourneyManagement/address"

File: extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json
Comment (cmathis) [inline extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:26]: Field name should be "https://ns.adobe.com/xdm/experience/customerJourneyManagement/operationalContext"

File: extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json
Comment (cmathis) [inline extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:8]: Remove the "xdm" from the $id path. Should be https://ns.adobe.com/experience/customerJourneyManagement/email-channel-context

**Comment by cmathis** [comment]
For all new mixins and fields use the "https://ns.adobe.com/xdm/experience/customerJourneyManagement/" namespace.

File: extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:20]: Hi @cmathis:  If we rename the field "https://ns.adobe.com/xdm/experience/customerJourneyManagement/address", then while ingesting or accessing the data, do we need to use the complete path, or it get stripped of while accessing the data through xed format?

cc: @aashishjassal

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.1.json
Comment (sawhneyh) [inline extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.1.json:15]: Added explicit measure field for building the reports in CJA. In absence of this explicit fields we have to use the calculated metrics for the send/bounce/click/open metrics.

File: extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:20]: @cmathis : Can you explain when we have to use the xdm namespace and when to use the complete uri path for defining the properties in mixin.

There are existing mixin like https://github.com/adobe/xdm/blob/master/extensions/adobe/experience/campaign/feedbackevent.schema.json which have properties defined using:

1. xdm namespance
2. uri style namespace
3. No namespaces.

Comment (cmathis) [reply extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:20]: Every field in XDM has a namespace. The URI style is the long form when you do not have a registered shorthand such as xdm, repo, adcloud, etc. There is no registered "customerJourneyManagement" short name so you would use the longer URI form which is fine as well. 

The path to this address field in AEP would become "_experience/customerJourneyManagmenet/address"

Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:20]: @cmathis : So if we are using the uri style namespaces for every field, and have a case like "domain" is a sub property of a "address" field. Then how will the path be mapped when we fetch the xdm given both "domain" and "address" have the same namespace path.
Do we have to access as: _experience/customerJourneyManagmenet/address. _experience/customerJourneyManagmenet/domain, or it will be _experience/customerJourneyManagmenet/address.domain as domain is a sub property of a address with same prefix namespace.

File: extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:8]: @cmathis : I noticed current test scripts do not validate the schemas under the "extension" folder. For validating the schemas with examples in this PR, I have to modify the test scripts locally to validate them. 
While running the schemas validating scripts it gives the error that "id" should match the following regex, and because of that I added the "xdm" in the uri path.  I will remove the "xdm" from the uri.

`(https://ns\.adobe\.com/xdm/[a-z0-9-/]*)|(http://schema\.org/.*)|(http://ns.adobe.com/adobecloud/core/1.0.*)|([http|https]://ns\.adobe\.com/adobecloudplatform/.*)|(https://tools\.ietf\.org/html/draft-kelly-json-hal-08/.*)|(http://www.iptc\.org/.*)|(https://id3\.org/id3v2.4/.*)`


On a side note, I assume this is a known issue, that current validation scripts do not validate schemas under the extension folder.

**Comment by sawhneyh** [comment]
@cmathis : You can continue the code review, I have incorporated the feedback to change the filed names to follow the uri pattern namespaces.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.1.json
Comment (aashishjassal) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.1.json:15]: +1

File: extensions/adobe/experience/customerJourneyManagement/messageprofile.schema.json
Comment (aashishjassal) [reply extensions/adobe/experience/customerJourneyManagement/messageprofile.schema.json:31]: As far as I also know there is a variant at a message profile level, complex nesting is there. @jfait Are you aware of any such possible requirements?

File: extensions/adobe/experience/customerJourneyManagement/push-channel-context.example.1.json
Comment (aashishjassal) [reply extensions/adobe/experience/customerJourneyManagement/push-channel-context.example.1.json:1]: There is an ambiguity on Mobile team right now. We are waiting for some discussion to finalize and will then float it to them

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (cmathis) [inline extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:33]: oneOf is not usable in AEP - you cannot have a field be two different data types.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (cmathis) [inline extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:97]: Mixins do not include the class in their allOf definition. The class will be included in the /schemas that is created along with this mixin.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (cmathis) [inline extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:100]: It appears you are including other mixins with this current definition. This is okay, but wanted to make sure that is what you wanted.

**Comment by prabhum2** [comment]
@sawhneyh - It appears that there is more internal discussion going on this . Please let me know when you are ready for this PR is ready to get into master

File: extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json
Comment (cmathis) [inline extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:20]: Need to remove the "xdm" from your URI string. Namespace should be https://ns.adobe.com/experience/customerJourneyManagement/. This is my fault because I had XDM in the example I showed in another comment.

File: extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json
Comment (cmathis) [reply extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json]: No the XED converter in AEP will recognize that the parent namespace is the same as the child namespace and will note require the duplication. It will be the second example - _experience/customerJourneyManagmenet/address.domain

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json]: @cmathis : removed the one of construct from here, but still using under the "required" properties to apply a constraint that "channelContext" can have either "emailChannelContext" or "pushChannelContext"

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:104]: @cmathis : We plan to have "message-experienceevent.schema" as a schema rather a mixin. Is it possible to define and commit pre built schemas like "message-experienceevent.schema".

Also similarly we have used mixin within mixin at line: https://github.com/adobe/xdm/blob/56573ad19fa756030ade3494552c88f3c01543a6/extensions/adobe/experience/customerJourneyManagement/messageprofile.schema.json#L25. I assume, we support adding mixin within mixin.

**Comment by prabhum2** [comment]
@sawhneyh - Can you please resolve above conflict

**Comment by sawhneyh** [comment]
@prabhum2 , @cmathis : You can resume the review, I have resolved the merged conflicts, and incorporated the review comments.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (cmathis) [inline extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:45]: I'm checking, but pretty sure the AEP ingestion process will not support this. "oneOf" is a json-schema concept not really supported in AEP at this time.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (cmathis) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:101]: So you want this to be a "global" cross tenant schema. This is technically possible, but not 100% supported in AEP. A user will not see this schema in the UI for example. Also not sure what will happen if they click on the schema of the Dataset. 

You can certainly try it though. To do that remove the "meta:intendedToExtend" (this is what makes it a mixin) and set the "meta:abstract" attribute to false.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (cmathis) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:104]: You can have a mixin which reference other mixins so this is fine.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (cmathis) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:45]: I checked and have confirmed that this oneOf around the "required" attribute is not supported by the ingestion process. There is no way today to say you are required to have field X OR Y. If this is a problem, then please open a ticket with the data ingestion team.

Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json]: Thanks @cmathis for update, I have removed the use of "oneOf" construct for defining the required properties.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:80]: @cmathis : Also can you confirm first part of the query:

>  We plan to have "message-experienceevent.schema" as a schema rather a mixin. Is it possible to define and commit pre built schemas like "message-experienceevent.schema".

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json]: @cmathis : You can resume the review. Also please provide your inputs regarding comment https://github.com/adobe/xdm/pull/915#discussion_r469872849.

We wanted to commit "message-experienceevent.schema" as a schema(not mixin which needs to be added over EE class) so that it can be available as OOTB schema in AEP.

If we have to make any changes for that, we will incorporate that.

**Comment by cmathis** [comment]
@sawhneyh  - Let us know if there is still internal conversations going on with this PR. Other then the build failure, I think this PR looks good. Just waiting for your approval that it is ready to be officially merged in.

File: extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json
Comment (cmathis) [reply extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json]: Yes, this is technically possible, but please schedule a meeting with XDM PM and myself to go over the details first.

File: extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json
Comment (sawhneyh) [inline extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json:20]: @cmathis : As per the https://github.com/adobe/xdm/blob/master/CONTRIBUTING.md we should not be inventing the new "id" attributes, but to use the "@id" convention.

Can you help usunderstand, how this "@id" field is namespaced, we may have a case where we add more than one mixin to schema, where both the mixjns has "@id" field. Then how this two "@id" fields are distinguished.

cc: @aashishjassal

File: extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json
Comment (cmathis) [inline extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:34]: oneOf not supported in AEP. This will throw an error when you try to use the schema there.

File: extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json
Comment (cmathis) [reply extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json:20]: The ExperienceEvent already includes an @id field, so this is basically ignored here. If you need an execution Id or something like it, then create a new field here named "executionId".

Comment (sawhneyh) [reply extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json]: I have replaced the "@id" field  with the full name spaced field name.

**Comment by sawhneyh** [comment]
1. I have fixed the lint issue.
2. Replaced the "@id" field with the proper name spaced field name.
3. Merged the email-channel-context and push-channel-context in messageProfile mixin.

@aashishjassal : Please review the changes.

**Review by aashishjassal — APPROVED** [review]

**Comment by aashishjassal** [comment]
Its approved, Please merge

**Comment by sawhneyh** [comment]
@cmathis : can you please merge this PR. Also once we merge this PR, will these mixins automatically starts reflecting in AEP, or we need to raise some request for that.
Also will mixins be first deployed on AEP int environment or on all environment at once.

**Review by cmathis — APPROVED** [review]

**Comment by cmathis** [comment]
@sawhneyh  - These changes are currently installed on INT. They will be pushed to Stage on Monday and stay there for one week before they go to Prod. Please test things on INT and Stage next week.

**Pull request merged**
